Fix Cassandra, improve config and tests

JohnAZoidberg opened the pull request:
###### Motivation for this change
- It's more intuitive and discoverable when you can configure set some options using Nix attributes - plus we can check it using some assertions.
- Make it possible to set the variables in `cassandra-env.sh`.
- Allow client connections by default because that's what users expect.
- Fix tests and test more things
- Make binaries work again (properly wrap them)

The test doesn't work - it doesn't work without this PR. I'll figure out why, when I have time. Seems like a race condition of the VMs starting, it's not a problem with the module itself, rather the tests.
[EDIT]: I think I fixed the tests, please have ofborg run them.

I tested manually on my system that the different configuration options actually have the desired effect.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

JohnAZoidberg committed:
nixos/cassandra: Enable CQL server by default

Resolves #50954

JohnAZoidberg committed:
nixos/cassandra: Allow setting of seed addresses

Allow for more intuitive specifying of seed node addresses with Nix
syntax.

JohnAZoidberg committed:
nixos/cassandra: Add nixos conf for Java env

JohnAZoidberg committed:
nixos/cassandra: Add clusterName

JohnAZoidberg committed:
nixos/cassandra: Fix test by listening on IP

Seems like you can't have a node as its own seed when it's listening on
an interface instead of an IP. At least the way it was done in the
test doesn't work and I can't figure out any other way than to just
listen on the IP address instead.

JohnAZoidberg commented:
This PR now also fixes the tests for Cassandra.
They take 5 minutes to run (on my laptop) but that's fine for booting and shutting down 3 VMs and rebalancing a Cassandra cluster.

JohnAZoidberg committed:
cassandra: Properly wrap all binaries

Would previously overwrite the binary with the wrapper and thus wrap
itself (resulting in an infinite recursion on execution) for the
binaries in /bin.t

jpyen commented:
This is very necessary. Thanks Daniel.

JohnAZoidberg commented:
I broke some of the binaries in #59090...
I misinterpreted the error when testing some of the binaries. I thought I just used it wrong parameters but turns out I broke some of them while making others work.
5041afc9690894f1a809ad6fc fixes it.

JohnAZoidberg commented:
Please somebody have a look - Cassandra is currently broken on master.

aanderse commented:
Hi John,

One delay here is that there is no clear indication of who the cassandra expert is for NixOS. Looking at the git history for both the package and service is a bit sparse. If no one is an expert on cassandra it will take a fair bit of work to review and validate these changes. So the question is... who do we need to get involved for review of this PR before someone with the appropriate access will feel comfortable merging?

aanderse commented:
@GrahamcOfBorg build cassandra
@GrahamcOfBorg test cassandra

aanderse commented:
@JohnAZoidberg cassandra is missing from all-tests.nix.

JohnAZoidberg committed:
nixos/cassandra: Reenable tests

JohnAZoidberg commented:
Actually yeah, that makes me think I should add my new config options to the tests so that non-Cassandra people can trust that they do the right thing. (And also they're automatically checked.)

JohnAZoidberg commented:
Does anybody know whether I can rely on the IPs staying like they are? Or can I force them to be like I assume?

aanderse commented:
> Does anybody know whether I can rely on the IPs staying like they are? Or can I force them to be like I assume?

Looking through `nixos/tests` I'm inclined to believe setting a static IP might be a good solution.

JohnAZoidberg committed:
nixos/cassandra: Allow setting JMX credentials

If we have the ability to enable remote JMX we should also support
setting credentials for that because they become required if you turn it
on.

JohnAZoidberg committed:
nixos/cassandra: Use ipAddress and capital letters

JohnAZoidberg committed:
nixos/cassandra: Test clusterName

JohnAZoidberg committed:
nixos/cassandra: Test seedAddresses

JohnAZoidberg committed:
nixos/cassandra: Set test VM IPs statically

JohnAZoidberg committed:
nixos/cassandra: Test maxHeapSize

JohnAZoidberg committed:
nixos/cassandra: Test jmxRoles

JohnAZoidberg commented:
I added some more tests that test almost all config options.

aanderse commented on nixos/tests/all-tests.nix:
```diff
@@ -35,6 +35,7 @@ in
   borgbackup = handleTest ./borgbackup.nix {};
   buildbot = handleTest ./buildbot.nix {};
   cadvisor = handleTestOn ["x86_64-linux"] ./cadvisor.nix {};
+  cassandra = handleTestOn ./cassandra.nix {};
```
Do you want `handleTest` or `handleTestOn` [ "..." ]?

JohnAZoidberg replied:
Ooops, thanks!

JohnAZoidberg commented:
@fuzzy-id You rewrote the module - do you have any thoughts?

aanderse commented:
Looks like openjdk on arch is broken in master :frowning_face: 
```
hash mismatch in fixed-output derivation '/nix/store/pc2krcggyflqbgpz6zlpzxr7hjyfdlif-OpenJDK11U-jre_aarch64_linux_hotspot_11.0.3_7.tar.gz':
  wanted: sha256:0hisjzgv29x2af0wvpy1c9j027qvv08cd9r9f0hg40wdczbbgxaa
  got:    sha256:0r45f358ih5xmygqm91g6nkq9d6nnjl8zz75kl4xbij00svzlcfy
```
Unrelated to this change, so I'm going to ignore those aarch64-linux errors.

fuzzy-id commented:
On first sight this looks good and promising. @JohnAZoidberg did you run the tests with all versions of cassandra we currently have in nixpkgs?

JohnAZoidberg commented:
No... Good point. It seems that the older versions don't work.
2.1 doesn't work because the password file has the wrong permissions and this version checks that - can we set permissions in the nix-store?
2.2 and 3.0 error with `sh: cannot set terminal process group (-1): Inappropriate ioctl for device` so I don't think it's the modules fault.

aanderse commented:
Please don't store passwords in the nix store! And no you cannot set permissions on files in the store, the nix store is *always* world readable!

JohnAZoidberg committed:
nixos/cassandra: Add option for password file path

If you're on a multi user system you don't want to have the password in
the nix-store. With the new jmxRolesFile option you can specify your own
protected file instead.

JohnAZoidberg committed:
nixos/cassandra: Use docbook instead of markdown style

JohnAZoidberg committed:
nixos/cassandra: Test remote auth only on v3.11+

Below that it works but only when supplying a custom password file with
restricted permissions (i.e. outside the nix-store). We can't do that
using an absolute path in the tests.

JohnAZoidberg commented:
Yeah, right - I knew that but I guess I was confused.

I added a warning not to use that on multi-user systems and added an option for providing a custom file.
I still kept it because in many cases you don't care and it's just easier to not have to keep track of a stateful file.

Tests work on all versions that we have in nixpkgs now - I verified that all versions we currently have are the ones supported by Apache.
The other error was unrelated and doesn't seem to cause any problems.

JohnAZoidberg commented:
Are you okay with that password config?
Can you please have ofBorg run the test?

aanderse commented:
@JohnAZoidberg with all the awesome work you've been doing lately you should request access to use the  bot yourself too! https://github.com/NixOS/ofborg

JohnAZoidberg commented:
Thanks! Did that yesterday ;)

aanderse commented:
@jpyen & @fuzzy-id as people with cassandra knowledge do you both approve the changes in this PR? Does everything seem reasonable in the changes? How about the test? Does the nixos test have good coverage? If you can provide in depth review it would be greatly appreciated.

@JohnAZoidberg looking over your (currently) 17 commits the only thing that jumps out is the massive number of options this module has. [RFC0042](https://github.com/NixOS/rfcs/pull/42) hasn't been accepted yet, and the reasoning you have for going this route seemed reasonable so the number of options can't really be held against you.

If the tests succeed, we get some reviews by people with cassandra knowledge, and no one else has any arguments against then I think we can merge this. This PR has sat for 24 days already with master being broken so hopefully we can wrap this up in the next few days. Thanks for all your hard work on nixpkgs lately :+1:

JohnAZoidberg commented:
Seems like ofBorg timed out after 30 minutes but the tests being run for only about three. On my laptops it takes 6 minutes - after all it starts a cluster with 2 nodes, breaks one and starts a third one to replace it. Cassandra has to do a lot of things.
I'm not sure what to do.

Yeah, I've seen the PR. I haven't thought about it a lot yet.
But I don't think it quite fully applies to Cassandra because here we use two config files, many optional systemd units and lots of environment variables.
I did add some more options which are needed/recommend for a minimal Cassandra cluster - I don't have the intention of adding all of the *hundreds* of supported options. See: 16dda79 (I should probably add an URL of the [docs](https://docs.datastax.com/en/cassandra/3.0/cassandra/configuration/configCassandra_yaml.html#configCassandra_yaml__QuickStartProps))

I split it up into different commits on purpose so that it's exactly clear what each of them does. Otherwise any reviewer would be terribly overwhelmed.

aanderse commented:
I just ran the test locally and had success. I understand your reasoning on the options. You did a great job of making the commits digestible for a reviewer, much appreciated. At this point I just want to hear someone else with cassandra knowledge agree with the changes in this PR and state that they feel the test is representative+adequate.

JohnAZoidberg committed:
nixos/cassandra: Don't force SimpleSeedProvider

If the `seedAddresses` is not set, don't force `SimpleSeedProvider` to
be in `seed_provider`. This could cause problems in a multi-datacenter
deployment when a different seed provider is preferred.

fuzzy-id reviewed (commented):
This looks strange to mee – `start_native_transport` is `true` by default when not set in the config. So this can actually only be used to disable native transport. But in the docs you reference options which then aren't available in the module options, i.e. `rpc_port` and `native_transport_port`. I doubt that this is helpful. I'd suggest to either drop this commit or provide the additional options favoring the former.

fuzzy-id reviewed (approved):
Good! According to documentation using a different `class_name` is usually not needed. Hence this provides good configurability without bloat.

JohnAZoidberg commented:
Yup, `start_native_transport` is enabled by default so that clients can connect to Cassandra when you `.enable = true` it. Would be confusing otherwise. See: #50954

What do you think about this diff? Is the doc clearer?
```diff
   cassandraConfig = flip recursiveUpdate cfg.extraConfig
     ({ commitlog_sync = "batch";
        commitlog_sync_batch_window_in_ms = 2;
-       start_native_transport = cfg.allowCql;
+       start_native_transport = cfg.allowClients;
        cluster_name = cfg.clusterName;
        listen_interface_prefer_ipv6 = cfg.listenInterfacePreferIPv6;           
        partitioner = "org.apache.cassandra.dht.Murmur3Partitioner";            
@@ -212,13 +212,16 @@ in {
         Set to 127.0.0.1 for a single node cluster.
       '';
     };
-    allowCql = mkOption {
+    allowClients = mkOption {
       type = types.bool;
       default = true;
       description = ''
         Enables or disables the native transport server (CQL binary protocol). 
-        This server uses the same address as the rpc_address, but the port it  
-        uses is not rpc_port but native_transport_port.
+        This server uses the same address as the <literal>rpcAddress</literal>,
+        but the port it uses is not <literal>rpc_port</literal> but            
+        <literal>native_transport_port</literal>. See the official Cassandra   
+        docs for more information on these variables and set them using        
+        <literal>extraConfig</literal>.
       '';
     };
     extraConfig = mkOption {
```


Yes, I didn't add them because they're very trivially addable using `extraConfig` and I want to add only the most important options in Nix.

fuzzy-id reviewed (approved):
Good! Tuning JVM options were needed very much.

fuzzy-id reviewed (changes requested):
Having `cluster_name` seems useful. Not sure about `listen_interface_prefer_ipv6` this looks too specific to me.

JohnAZoidberg commented:
> Not sure about `listen_interface_prefer_ipv6` this looks too specific to me.

You're right, it does. But DataStax include it in their "minimal properties needed for configuring a cluster" :shrug:
And a bit IPv6 propaganda can't hurt ;)

fuzzy-id commented:
> Yup, `start_native_transport` is enabled by default so that clients can connect to Cassandra when you `.enable = true` it. Would be confusing otherwise. See: #50954

This is strange. I was referring to the [Cassandra docs](https://cassandra.apache.org/doc/latest/configuration/cassandra_config_file.html#start-native-transport) which state that `start_native_transport` is `true` by default. Hence this option should be superfluous. BUT I just tested locally and noticed that it is indeed necessary. o_O Looks like this is a documentation bug… 

> 
> What do you think about this diff? Is the doc clearer?
> […]

Yep. This looks very good to me.

fuzzy-id commented:
I'm not convinced of the `listen_interface_prefer_ipv6` option. It only makes sense if `listen_interface` is set, too. (But having an addiitonal test for this corner-case doesn't seem right neither.) Also note the [docs](https://cassandra.apache.org/doc/latest/configuration/cassandra_config_file.html#listen-interface-prefer-ipv6): 

> If there is only one address it will be selected regardless of ipv4/ipv6.

Maybe you want to add this to the documentation to produce less surprises. In general, this option seems to have a sensible default, so I don't see the necessity to have it here.

All the rest looks very, very good to me. Thanks a lot! Tests look solid. The module itself looks cleaner. I especially appreciate the addiitonal possibility to configure JVM options.

JohnAZoidberg commented:
> I'm not convinced of the `listen_interface_prefer_ipv6` option. It only makes sense if `listen_interface` is set, too. (But having an additional test for this corner-case doesn't seem right neither.) Also note the [docs](https://cassandra.apache.org/doc/latest/configuration/cassandra_config_file.html#listen-interface-prefer-ipv6):

Yeah okay, I removed it. Like I previously said: We don't want to include every possible config in Nix.

> This is strange. I was referring to the Cassandra docs which state that start_native_transport is true by default. Hence this option should be superfluous. BUT I just tested locally and noticed that it is indeed necessary. o_O Looks like this is a documentation bug…

I've wondered too... Yes, seems like the doc is wrong.

Thanks guys for the patience and thorough reviews :)

fuzzy-id commented:
> > I'm not convinced of the `listen_interface_prefer_ipv6` option. […]
> 
> Yeah okay, I removed it. Like I previously said: We don't want to include every possible config in Nix.

Perfect! I think this module makes a good compromise on exposed options now.

> Thanks guys for the patience and thorough reviews :)

Dito. Good work!

From my point of view this is ready to get merged.

aanderse commented:
If either @jpyen or @contrun can weigh in as a third set of eyes/final review I would greatly appreciate.

aanderse commented:
@JohnAZoidberg This has sat for long enough. We're not getting any more review than we've already had. Do you mind fixing the conflict and then we'll merge? Thanks for your work and patience on this.

JohnAZoidberg commented:
Oh, I was responsible for those conflicts :laughing:

aanderse commented:
Thanks @JohnAZoidberg

ivan commented:
The NixOS manual doesn't seem to build with this merged:

```
# nixos-rebuild switch --builders ''
building Nix...
building the system configuration...
trace: warning: Option `programs.captive-browser.package' has no description.
these derivations will be built:
  /nix/store/d9ihpqcd0d6x5dswjfkwrsiwrv1jd2gk-options-db.xml.drv
  /nix/store/33p1av88g1y974qgjbmxz3a53nsv61vp-generated-docbook.drv
  /nix/store/w7zbd8q8akfdcy47vwplpz8n63kwi61b-nixos-manual-combined.drv
  /nix/store/y8z3qqwv7ivskf4m8rzqiz6jpqrgmk1k-manual-olinkdb.drv
  /nix/store/aslfagf5y3c267g41aj2kr3c9j2z8zvm-nixos-manual-html.drv
  /nix/store/dfsssp62vf9963yqdcq9wwc5axljxki2-nixos-help.drv
  /nix/store/rjpb8pnpk4m30dwbi245fpz4bycqb9x0-nixos-manual.desktop.drv
  /nix/store/yfkra26ay0lc76j06cbypj9dmk4c677l-nixos-manpages.drv
  /nix/store/y3qm2sy4y3czpisxdfn67g8vnpn5ccjl-system-path.drv
  /nix/store/n2k8cpgsmzs61v3jhfq4g1v694q1kyjp-dbus-1.drv
  /nix/store/sr7ywq644wvzcdaqm82mha8kf6nf0rw0-unit-dbus.service.drv
  /nix/store/h9dnqlg1pbkhnwx4040cff2z3bx4ab24-user-units.drv
  /nix/store/lija5f5blhlm19clnvlr8l98zpwh4clj-unit-polkit.service.drv
  /nix/store/p7xypymw1czr1rmkiy65rd8ki2xgy4bi-unit-dbus.service.drv
  /nix/store/wqzj1ycaw5va2iy3ddsjdbxckgb01hxj-unit-systemd-fsck-.service.drv
  /nix/store/wrig2rm34mbwq3sfyz5zpzii5xq04hda-system-units.drv
  /nix/store/w31lswqzbb21hgn4hy5inyq697kd7kw8-etc.drv
  /nix/store/0n3zi29zf8k7p1l46qkxy3qn0yh69qh0-nixos-system-ra-19.09.git.412d110.drv
building '/nix/store/d9ihpqcd0d6x5dswjfkwrsiwrv1jd2gk-options-db.xml.drv'...
intermediate.xml:12476: parser error : Opening and ending tag mismatch: literal line 12471 and para
</para></nixos:option-description><para><emphasis>Type:</emphasis> null or strin
       ^
intermediate.xml:12476: parser error : Opening and ending tag mismatch: literal line 12471 and nixos:option-description
</para></nixos:option-description><para><emphasis>Type:</emphasis> null or strin
                                  ^
intermediate.xml:12480: parser error : Opening and ending tag mismatch: para line 12470 and listitem
            </filename></member></simplelist></listitem></varlistentry><varliste
                                                        ^
intermediate.xml:12480: parser error : Opening and ending tag mismatch: option-description line 12470 and varlistentry
            </filename></member></simplelist></listitem></varlistentry><varliste
                                                                       ^
intermediate.xml:60118: parser error : Opening and ending tag mismatch: listitem line 12470 and variablelist
      </filename></member></simplelist></listitem></varlistentry></variablelist>
                                                                               ^
intermediate.xml:60118: parser error : Opening and ending tag mismatch: varlistentry line 12470 and appendix
lename></member></simplelist></listitem></varlistentry></variablelist></appendix
                                                                               ^
intermediate.xml:60120: parser error : Premature end of data in tag variablelist line 3

^
intermediate.xml:60120: parser error : Premature end of data in tag appendix line 3

^
unable to parse intermediate.xml
builder for '/nix/store/d9ihpqcd0d6x5dswjfkwrsiwrv1jd2gk-options-db.xml.drv' failed with exit code 6
cannot build derivation '/nix/store/33p1av88g1y974qgjbmxz3a53nsv61vp-generated-docbook.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/w7zbd8q8akfdcy47vwplpz8n63kwi61b-nixos-manual-combined.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/yfkra26ay0lc76j06cbypj9dmk4c677l-nixos-manpages.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/aslfagf5y3c267g41aj2kr3c9j2z8zvm-nixos-manual-html.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/y3qm2sy4y3czpisxdfn67g8vnpn5ccjl-system-path.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/0n3zi29zf8k7p1l46qkxy3qn0yh69qh0-nixos-system-ra-19.09.git.412d110.drv': 1 dependencies couldn't be built
error: build of '/nix/store/0n3zi29zf8k7p1l46qkxy3qn0yh69qh0-nixos-system-ra-19.09.git.412d110.drv' failed
```

turion commented on nixos/modules/services/databases/cassandra.nix:
```diff
@@ -39,15 +42,42 @@ let
+        cp "$cassandraEnvPkg" "$out/cassandra-env.sh"
+
+        # Delete default JMX Port, otherwise we can't set it using env variable
+        sed -i '/JMX_PORT="7199"/d' "$out/cassandra-env.sh"
```
If the source line ever changes, this will just fail silently?

JohnAZoidberg replied:
If they remove or change this line, sed will not remove it and not issue a warning.
`substituteInPlace` would warn, but it doesn't have an option to delete an entire line.

I think it'll be fine if I implement the test for `jmxPort` like you suggested below.

turion commented on nixos/tests/cassandra.nix:
```diff
@@ -1,67 +1,109 @@
-      networking.firewall.enable = false;
-      services.cassandra = cassandraCfg // extra;
+      networking = {
+        firewall.allowedTCPPorts = [ 7000 7199 9042 ];
```
Instead of `7199`, don't you want `jmxPort`?

JohnAZoidberg replied:
Good idea, see #65616

turion reviewed (commented):
Just some minor improvement comments, but I can't say I understand nix well enough to judge completely.
